src/qibo/parameter.py

```diff
@@ -0,0 +1,134 @@
+        nofeatures (bool): flag to explicitly ban the updating of the features. This simplifies the task of updating Parameter objects simultaneously when some have embedded features and some do not.
+    """
+
+    def __init__(self, func, trainable=[], features=[]):
+        self.trainable = trainable
+        self.features = features
+
+        if self.nfeat + self.nparams != func.__code__.co_argcount:
+            raise_error(
+                TypeError,
+                f"{self.nfeat + self.nparams} parameters are provided, but {func.__code__.co_argcount} are required, please initialize features and trainable according to the defined function.",
+            )
+        # lambda function
+        self.lambdaf = func
+
+        # calculate derivatives
+        # maybe here use JAX ?
+        self.derivatives = calculate_derivatives(func=self.lambdaf)
+
+    def __call__(self, features=None, trainable=None):
+        """Return parameter value with given features and/or trainable."""
+
+        params = []
+
+        if features is None:
+            params.extend(self.features)
+        else:
+            if len(features) != self.nfeat:
+                raise_error(
+                    TypeError,
+                    f"The number of features provided is not compatible with the problem's dimensionality, which is {self.nfeat}.",
+                )
+            else:
+                params.extend(features)
+        if trainable is None:
+            params.extend(self.trainable)
+        else:
+            if len(trainable) != self.nparams:
+                raise_error(
+                    TypeError,
+                    f"The number of trainable provided is different from the number of required parameters, which is {self.nparams}.",
+                )
+            else:
+                params.extend(trainable)
+
+        return self.lambdaf(*params)
+
+    @property
+    def nparams(self):
+        """Returns the number of trainable parameters"""
+        try:
+            return len(self.trainable)
+        except TypeError:
+            return 0
+
+    @property
+    def nfeat(self):
+        """Returns the number of features"""
+        return len(self.features) if isinstance(self.features, list) else 0
+
+    @property
+    def ncomponents(self):
+        """Returns the number of elements which compose the Parameter"""
+        return self.nparams + self.nfeat
+
+    def trainable_parameter_indices(self, start_index):
+        """Return list of respective indices of trainable parameters within
+        the larger trainable parameter list of a circuit for example"""
+        return (np.arange(self.nparams) + start_index).tolist()
+
+    def unaffected_by(self, trainable_idx):
+        """Retrieve constant term of lambda function with regard to a specific trainable parameter"""
+        params = self.trainable.copy()
+        params[trainable_idx] = 0.0
+        return self.__call__(trainable=params)
```

Comment:
```suggestion
        return self(trainable=params)
```

They are aliases of each other. If you can use the public, prefer it over the private (in particular over magic methods).

In case you could not use `self(...)`, even better to use:
```suggestion
        import operator
        return operator.call(self, trainable=params)
```